<a href="https://colab.research.google.com/github/AlbericoJose/Coleta-e-Analise/blob/main/Coleta_dados_vacina.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Coleta de dados estruturados

In [ ]:
import pandas as pd
import os

# Importando dados do arquivo


In [ ]:
DIRETORIO = 'C:\\Projetos\\coleta_dados_estruturados\\datasets\\'
arquivo = 'country_vaccinations.csv'
dados = os.path.join(DIRETORIO, arquivo) # busca o caminho do arquivo já com as converções

# Verifica o caminho do diretório que o arquivo esta salvo

In [ ]:
dados

'C:\\Projetos\\coleta_dados_estruturados\\datasets\\country_vaccinations.csv'

#  Entrar no diretório onde está o arquivo e transformar em um dataframe do pandas

In [ ]:
dataframe = pd.read_csv(dados, sep=',', encoding='latin-1')

In [ ]:
dataframe.head(5)

country iso_code        date  total_vaccinations  people_vaccinated  \
0  Afghanistan      AFG  2021-02-22                 0.0                0.0   
1  Afghanistan      AFG  2021-02-23                 NaN                NaN   
2  Afghanistan      AFG  2021-02-24                 NaN                NaN   
3  Afghanistan      AFG  2021-02-25                 NaN                NaN   
4  Afghanistan      AFG  2021-02-26                 NaN                NaN   

   people_fully_vaccinated  daily_vaccinations_raw  daily_vaccinations  \
0                      NaN                     NaN                 NaN   
1                      NaN                     NaN              1367.0   
2                      NaN                     NaN              1367.0   
3                      NaN                     NaN              1367.0   
4                      NaN                     NaN              1367.0   

   total_vaccinations_per_hundred  people_vaccinated_per_hundred  \
0                             0.0                            0.0   
1                             NaN                            NaN   
2                             NaN                            NaN   
3                             NaN                            NaN   
4                             NaN                            NaN   

   people_fully_vaccinated_per_hundred  daily_vaccinations_per_million  \
0                                  NaN                             NaN   
1                                  NaN                            34.0   
2                                  NaN                            34.0   
3                                  NaN                            34.0   
4                                  NaN                            34.0   

                                            vaccines  \
0  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
1  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
2  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
3  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
4  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   

                 source_name            source_website  
0  World Health Organization  https://covid19.who.int/  
1  World Health Organization  https://covid19.who.int/  
2  World Health Organization  https://covid19.who.int/  
3  World Health Organization  https://covid19.who.int/  
4  World Health Organization  https://covid19.who.int/

# Analisando os dados

In [ ]:
#dados referente ao Brasil
dataframe.query('country == "Brazil"').sort_values('date', ascending=False)[:5]

country iso_code        date  total_vaccinations  people_vaccinated  \
11085  Brazil      BRA  2022-03-29         413559595.0        181078067.0   
11084  Brazil      BRA  2022-03-28         412177025.0        180881628.0   
11083  Brazil      BRA  2022-03-27         411548202.0        180800024.0   
11082  Brazil      BRA  2022-03-26         411116197.0        180757995.0   
11081  Brazil      BRA  2022-03-25         410860641.0        180726798.0   

       people_fully_vaccinated  daily_vaccinations_raw  daily_vaccinations  \
11085              160272858.0               1382570.0            656803.0   
11084              160008924.0                628823.0            546999.0   
11083              159842858.0                432005.0            587488.0   
11082              159756178.0                255556.0            548819.0   
11081              159673499.0                684805.0            596456.0   

       total_vaccinations_per_hundred  people_vaccinated_per_hundred  \
11085                          193.26                          84.62   
11084                          192.61                          84.53   
11083                          192.32                          84.49   
11082                          192.12                          84.47   
11081                          192.00                          84.45   

       people_fully_vaccinated_per_hundred  daily_vaccinations_per_million  \
11085                                74.90                          3069.0   
11084                                74.77                          2556.0   
11083                                74.70                          2745.0   
11082                                74.65                          2565.0   
11081                                74.62                          2787.0   

                                                vaccines  \
11085  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
11084  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
11083  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
11082  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
11081  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   

                                           source_name  \
11085  State governments via coronavirusbra1.github.io   
11084  State governments via coronavirusbra1.github.io   
11083  State governments via coronavirusbra1.github.io   
11082  State governments via coronavirusbra1.github.io   
11081  State governments via coronavirusbra1.github.io   

                          source_website  
11085  https://coronavirusbra1.github.io  
11084  https://coronavirusbra1.github.io  
11083  https://coronavirusbra1.github.io  
11082  https://coronavirusbra1.github.io  
11081  https://coronavirusbra1.github.io

In [ ]:
# numero de paises distintos na base
dataframe['country'].nunique()

223

# Qual o país com maior número de pessoas imunizadas ?

In [ ]:
# identificando a data mais recente da coleta
ultima_data = dataframe['date'].max()
ultima_data

'2022-03-29'

In [ ]:
# filtrando a última data
ultima_data = dataframe[dataframe.date == ultima_data]

In [ ]:
# quantidade de linhas e colunas dentro do dataset
ultima_data.shape

(54, 15)

# Atenção
Como temos 223 paíes e quando buscamos a data mais recente, só aparecem 54, teremos que fazer outro tipo de filtro no dataset, pois, acabaremos excluindo dados de ourtos países.

In [ ]:
dataframe.groupby('country')['date'].max().sort_values(ascending=False)

country
Lebanon                            2022-03-29
Lithuania                          2022-03-29
Kosovo                             2022-03-29
Kazakhstan                         2022-03-29
Japan                              2022-03-29
                                      ...    
Pitcairn                           2021-09-07
Turkmenistan                       2021-09-05
Bonaire Sint Eustatius and Saba    2021-09-01
Saint Helena                       2021-05-05
Falkland Islands                   2021-04-14
Name: date, Length: 223, dtype: object

In [ ]:
dataframe.groupby('country')['total_vaccinations'].max().sort_values(ascending=False)

country
China               3.263129e+09
India               1.834501e+09
United States       5.601818e+08
Brazil              4.135596e+08
Indonesia           3.771089e+08
                        ...     
Falkland Islands    4.407000e+03
Montserrat          4.211000e+03
Niue                4.161000e+03
Tokelau             1.936000e+03
Pitcairn            9.400000e+01
Name: total_vaccinations, Length: 223, dtype: float64

# Top 10 paises

In [ ]:
imunizados_por_pais = dataframe.groupby('country')['total_vaccinations'].max().sort_values(ascending=False)
imunizados_por_pais

country
China               3.263129e+09
India               1.834501e+09
United States       5.601818e+08
Brazil              4.135596e+08
Indonesia           3.771089e+08
                        ...     
Falkland Islands    4.407000e+03
Montserrat          4.211000e+03
Niue                4.161000e+03
Tokelau             1.936000e+03
Pitcairn            9.400000e+01
Name: total_vaccinations, Length: 223, dtype: float64

In [ ]:
type(imunizados_por_pais)

pandas.core.series.Series

In [ ]:
imunizados = imunizados_por_pais.to_frame().reset_index()
imunizados

country  total_vaccinations
0               China        3.263129e+09
1               India        1.834501e+09
2       United States        5.601818e+08
3              Brazil        4.135596e+08
4           Indonesia        3.771089e+08
..                ...                 ...
218  Falkland Islands        4.407000e+03
219        Montserrat        4.211000e+03
220              Niue        4.161000e+03
221           Tokelau        1.936000e+03
222          Pitcairn        9.400000e+01

[223 rows x 2 columns]

In [ ]:
imunizados.sort_values('total_vaccinations', ascending=False).head(10)

country  total_vaccinations
0          China        3.263129e+09
1          India        1.834501e+09
2  United States        5.601818e+08
3         Brazil        4.135596e+08
4      Indonesia        3.771089e+08
5          Japan        2.543456e+08
6     Bangladesh        2.436427e+08
7       Pakistan        2.193686e+08
8        Vietnam        2.031444e+08
9         Mexico        1.919079e+08

# Qual vacina é mais usada no mundo ?

In [ ]:
vacinas_por_pais = dataframe[['country', 'date', 'vaccines']]
vacinas_por_pais

country        date  \
0      Afghanistan  2021-02-22   
1      Afghanistan  2021-02-23   
2      Afghanistan  2021-02-24   
3      Afghanistan  2021-02-25   
4      Afghanistan  2021-02-26   
...            ...         ...   
86507     Zimbabwe  2022-03-25   
86508     Zimbabwe  2022-03-26   
86509     Zimbabwe  2022-03-27   
86510     Zimbabwe  2022-03-28   
86511     Zimbabwe  2022-03-29   

                                                vaccines  
0      Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...  
1      Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...  
2      Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...  
3      Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...  
4      Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...  
...                                                  ...  
86507  Oxford/AstraZeneca, Sinopharm/Beijing, Sinovac...  
86508  Oxford/AstraZeneca, Sinopharm/Beijing, Sinovac...  
86509  Oxford/AstraZeneca, Sinopharm/Beijing, Sinovac...  
86510  Oxford/AstraZeneca, Sinopharm/Beijing, Sinovac...  
86511  Oxford/AstraZeneca, Sinopharm/Beijing, Sinovac...  

[86512 rows x 3 columns]

Agrupar por vacina para descobrir quais vacinas são mais usadas em cada país

In [ ]:
vacinas_por_pais.groupby('vaccines')['country'].nunique().sort_values(ascending=False)

vaccines
Oxford/AstraZeneca                                                                                    20
Johnson&Johnson, Moderna, Oxford/AstraZeneca, Pfizer/BioNTech                                         17
Moderna, Oxford/AstraZeneca, Pfizer/BioNTech                                                          15
Oxford/AstraZeneca, Pfizer/BioNTech                                                                   11
Johnson&Johnson, Moderna, Novavax, Oxford/AstraZeneca, Pfizer/BioNTech                                 8
                                                                                                      ..
Moderna, Oxford/AstraZeneca, Pfizer/BioNTech, Sinopharm/Beijing, Sinovac                               1
Moderna, Oxford/AstraZeneca, Pfizer/BioNTech, Sinovac, Sputnik Light, Sputnik V, ZF2001                1
Moderna, Oxford/AstraZeneca, Pfizer/BioNTech, Sinovac, Sputnik V                                       1
Moderna, Oxford/AstraZeneca, Pfizer/BioNTech, 

In [ ]:
vacinas_pais = dataframe.groupby('country')['vaccines'].unique()
vacinas_pais = vacinas_pais.to_frame().reset_index()

# Qual o país possui o maior número de vacinados por dia ?

In [ ]:
dataframe.groupby('country')['daily_vaccinations'].max().sort_values(ascending=False)

country
China            22424286.0
India            10037995.0
Bangladesh        3758404.0
United States     3506960.0
Pakistan          2175773.0
                    ...    
Saint Helena          165.0
Niue                   87.0
Montserrat             53.0
Tokelau                23.0
Pitcairn                1.0
Name: daily_vaccinations, Length: 223, dtype: float64

# Qual a média de vacinação por país

In [ ]:
media_pais = dataframe.groupby('country')['daily_vaccinations'].mean().sort_values(ascending=False)
media_pais = media_pais.to_frame().reset_index()
media_pais

country  daily_vaccinations
0                China        6.930368e+06
1                India        4.175994e+06
2        United States        1.191727e+06
3               Brazil        9.435287e+05
4            Indonesia        8.462893e+05
..                 ...                 ...
218  Wallis and Futuna        3.388649e+01
219            Tokelau        1.739823e+01
220               Niue        1.631698e+01
221         Montserrat        1.053659e+01
222           Pitcairn        5.119048e-01

[223 rows x 2 columns]

dataframes criados

1.Imunizados

2.Vacinas_por_pais

3.media_pais

Com base nos dataframes criados vamos criar uma nova base de dados utilizando os dados que tratamos anteriormente

In [ ]:
# Renomeando novas colunas das datas
# vamos inicar alterando o nome de algumas colunas

media_pais = media_pais.rename(columns={'country' : 'pais', 'daily_vaccinations' : 'media_vacinacao'})

In [ ]:
media_pais

pais  media_vacinacao
0                China     6.930368e+06
1                India     4.175994e+06
2        United States     1.191727e+06
3               Brazil     9.435287e+05
4            Indonesia     8.462893e+05
..                 ...              ...
218  Wallis and Futuna     3.388649e+01
219            Tokelau     1.739823e+01
220               Niue     1.631698e+01
221         Montserrat     1.053659e+01
222           Pitcairn     5.119048e-01

[223 rows x 2 columns]

In [ ]:
vacinas_pais = vacinas_pais.rename(columns={'country' : 'pais', 'vaccines' : 'vacina_utilizada'})
vacinas_pais[:5]

pais                                   vacina_utilizada
0  Afghanistan  [Johnson&Johnson, Oxford/AstraZeneca, Pfizer/B...
1      Albania  [Oxford/AstraZeneca, Pfizer/BioNTech, Sinovac,...
2      Algeria  [Oxford/AstraZeneca, Sinopharm/Beijing, Sinova...
3      Andorra     [Moderna, Oxford/AstraZeneca, Pfizer/BioNTech]
4       Angola                               [Oxford/AstraZeneca]

In [ ]:
imunizados = imunizados.rename(columns={'country' : 'pais', 'total_vaccinations' : 'imunizados'})
imunizados[:5]

pais    imunizados
0          China  3.263129e+09
1          India  1.834501e+09
2  United States  5.601818e+08
3         Brazil  4.135596e+08
4      Indonesia  3.771089e+08

# Criando um novo dataset

In [ ]:
dados_tratados = pd.merge(imunizados, media_pais, how='inner')

In [ ]:
dados_tratados = pd.merge(dados_tratados, vacinas_pais)
dados_tratados

pais    imunizados  media_vacinacao  \
0               China  3.263129e+09     6.930368e+06   
1               India  1.834501e+09     4.175994e+06   
2       United States  5.601818e+08     1.191727e+06   
3              Brazil  4.135596e+08     9.435287e+05   
4           Indonesia  3.771089e+08     8.462893e+05   
..                ...           ...              ...   
218  Falkland Islands  4.407000e+03     7.474242e+01   
219        Montserrat  4.211000e+03     1.053659e+01   
220              Niue  4.161000e+03     1.631698e+01   
221           Tokelau  1.936000e+03     1.739823e+01   
222          Pitcairn  9.400000e+01     5.119048e-01   

                                      vacina_utilizada  
0    [CanSino, Sinopharm/Beijing, Sinopharm/Wuhan, ...  
1             [Covaxin, Oxford/AstraZeneca, Sputnik V]  
2          [Johnson&Johnson, Moderna, Pfizer/BioNTech]  
3    [Johnson&Johnson, Oxford/AstraZeneca, Pfizer/B...  
4    [Johnson&Johnson, Moderna, Novavax, Oxford/Ast...  
..                                                 ...  
218                               [Oxford/AstraZeneca]  
219                               [Oxford/AstraZeneca]  
220                                  [Pfizer/BioNTech]  
221                                  [Pfizer/BioNTech]  
222                               [Oxford/AstraZeneca]  

[223 rows x 4 columns]

# Exportando dados tratados

In [ ]:
dados_tratados.to_csv('estatistica.csv', index=True, sep=';')